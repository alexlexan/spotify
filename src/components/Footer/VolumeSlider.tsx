import React, { ChangeEvent } from "react";
import "./Footer.sass";
import VolumeDownIcon from "@material-ui/icons/VolumeDown";
import { Grid, Slider } from "@material-ui/core";
import { useSelector, useDispatch } from "react-redux";
import Sound from "react-sound";
import { stop } from "../../store/actions";
import { AppState } from "../../store/rootReducer";
import { State, Track } from "../../store/reducer";
import { Actions } from "../../store/actions";
import { ThunkDispatch } from "redux-thunk";

type VolumeSliderProps = {
  activeTrack: Track | null;
  playing: "PLAYING" | "STOPPED" | "PAUSED";
};

const VolumeSlider: React.FC<VolumeSliderProps> = ({
  activeTrack,
  playing,
}) => {
  const dispatch: ThunkDispatch<State, unknown, Actions> = useDispatch();
  const volume = useSelector((state: AppState) => state.spotify.volume);
  const oldVolume = React.useRef(volume);

  const finish = () => {
    dispatch(stop());
  };

  const handleSliderChange = (
    _: ChangeEvent<{}>,
    newValue: number | number[]
  ) => {
    dispatch({ type: "VOLUME", volume: newValue as number });
    oldVolume.current = newValue as number;
  };

  const muteVolume = (volume: number, oldVolume: number | number[]) => {
    const newVolume = volume === 0 ? oldVolume : 0;
    dispatch({ type: "VOLUME", volume: newVolume as number });
  };

  return (
    <Grid container spacing={2}>
      <Grid item>
        <VolumeDownIcon onClick={() => muteVolume(volume, oldVolume.current)} />
      </Grid>
      <Grid item xs>
        {activeTrack && (
          <Sound
            url={activeTrack.preview_url || activeTrack.external_urls.spotify}
            playStatus={playing}
            volume={volume}
            onFinishedPlaying={finish}
          />
        )}
        <Slider
          aria-labelledby="continuous-slider"
          value={volume}
          onChange={handleSliderChange}
        />
      </Grid>
    </Grid>
  );
};

export default VolumeSlider;
