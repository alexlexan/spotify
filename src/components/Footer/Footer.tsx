import React from "react";
import "./Footer.sass";
import PlayCircleOutlineIcon from "@material-ui/icons/PlayCircleOutline";
import SkipPreviousIcon from "@material-ui/icons/SkipPrevious";
import SkipNextIcon from "@material-ui/icons/SkipNext";
import PauseCircleOutlineIcon from "@material-ui/icons/PauseCircleOutline";
import { useDispatch, useSelector } from "react-redux";
import { play, pause, playAndSetTrack } from "../../store/actions";
import { AppState } from "../../store/rootReducer";
import { Track } from "../../store/reducer";
import VolumeSlider from "./VolumeSlider";

const Footer: React.FC = () => {
  const playing = useSelector((state: AppState) => state.spotify.playing);
  const playlists = useSelector(
    (state: AppState) => state.spotify.activePlaylist
  );
  const activeTrack = useSelector(
    (state: AppState) => state.spotify.activeTrack
  );
  const songs = useSelector((state: AppState) => state.spotify.songs);
  const dispatch = useDispatch();

  const nextTrack = (songs: Track[] | null) => {
    if (songs && activeTrack) {
      const nextItem = songs.findIndex((item) => activeTrack.id === item.id);
      songs[nextItem + 1] && dispatch(playAndSetTrack(songs[nextItem + 1]));
    }
  };

  const prevTrack = (songs: Track[] | null) => {
    if (songs && activeTrack) {
      const nextItem = songs.findIndex((item) => activeTrack.id === item.id);
      songs[nextItem - 1] && dispatch(playAndSetTrack(songs[nextItem - 1]));
    }
  };

  const pauseTrack = () => {
    activeTrack && dispatch(pause());
  };
  const playTrack = () => {
    activeTrack && dispatch(play());
  };

  return (
    <div className="footer">
      <div className="footer__left">
        {activeTrack && (
          <img
            className="footer__albumLogo"
            src={activeTrack?.album?.images[0].url || playlists?.images[0].url}
            alt={`${activeTrack?.name}`}
          />
        )}
        {activeTrack ? (
          <div className="footer__songInfo">
            <h4>{activeTrack.name}</h4>
            <p>{activeTrack.artists.map((artist) => artist.name).join(", ")}</p>
          </div>
        ) : (
          <div className="footer__songInfo">
            <h4>No song is playing</h4>
            <p>...</p>
          </div>
        )}
      </div>

      <div className="footer__center">
        <SkipPreviousIcon
          className="footer__icon"
          onClick={() => prevTrack(songs)}
        />
        {playing === "PLAYING" ? (
          <PauseCircleOutlineIcon
            onClick={pauseTrack}
            fontSize="large"
            className="footer__icon"
          />
        ) : (
          <PlayCircleOutlineIcon
            fontSize="large"
            className="footer__icon"
            onClick={playTrack}
          />
        )}
        <SkipNextIcon
          className="footer__icon"
          onClick={() => nextTrack(songs)}
        />
      </div>
      <div className="footer__right">
        <VolumeSlider activeTrack={activeTrack} playing={playing} />
      </div>
    </div>
  );
};

export default Footer;
