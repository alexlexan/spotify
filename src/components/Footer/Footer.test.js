import React from "react";
import Footer from "./Footer";
import { useSelector } from "react-redux";
import Sound from "react-sound";
import PlayCircleOutlineIcon from "@material-ui/icons/PlayCircleOutline";
import SkipPreviousIcon from "@material-ui/icons/SkipPrevious";
import SkipNextIcon from "@material-ui/icons/SkipNext";
import PauseCircleOutlineIcon from "@material-ui/icons/PauseCircleOutline";

const mockDispatch = jest.fn();
jest.mock("react-redux", () => ({
  useSelector: jest.fn(),
  useDispatch: () => mockDispatch,
}));

describe("test Footer component", () => {
  let component, songs, firstTrack, secondTrack, thirdTrack;
  firstTrack = {
    id: 1,
    name: "firstTrack",
    type: "Track",
    images: [{ url: "firstTrackUrl" }],
    artists: [{ name: "Artist" }],
  };
  secondTrack = {
    id: 2,
    name: "secondTrack",
    type: "Track",
    images: [{ url: "secondTrackUrl" }],
    artists: [{ name: "Artist" }],
  };
  thirdTrack = {
    id: 3,
    name: "thirdTrack",
    type: "Track",
    images: [{ url: "thirdTrack" }],
    artists: [{ name: "Artist" }],
    album: { images: [{ url: "urlalbumPlaylist" }] },
  };

  afterEach(() => {
    mockDispatch.mockClear();
  });
  describe("render Footer component without activeTrack, without playing", () => {
    songs = [firstTrack, secondTrack];

    beforeEach(() => {
      useSelector
        .mockReturnValueOnce(Sound.status.STOPPED)
        .mockReturnValueOnce({
          images: [{ url: "urlPlaylist" }],
          name: "playlist",
        })
        .mockReturnValueOnce(null)
        .mockReturnValueOnce(songs);
      component = shallow(<Footer />);
    });

    it("snapshot Footer component without activeTrack, without playing", () => {
      expect(component).toMatchSnapshot();
    });

    it("test click SkipPreviousIcon component should dispatch pause action", () => {
      component.find(SkipPreviousIcon).simulate("click");
      expect(mockDispatch).toBeCalledTimes(0);
    });

    it("test click SkipNextIcon component should dispatch pause action", () => {
      component.find(SkipNextIcon).simulate("click");
      expect(mockDispatch).toBeCalledTimes(0);
    });
  });

  describe("test", () => {
    songs = [firstTrack, secondTrack, thirdTrack];
    it("snapshot Footer component with activeTrack with activeTrack.album, without playing", () => {
      useSelector.mockImplementation((selector) =>
        selector({
          spotify: {
            playing: Sound.status.STOPPED,
            activePlaylist: {
              images: [{ url: "urlPlaylist" }],
              name: "playlist",
            },
            activeTrack: thirdTrack,
            songs: songs,
          },
        })
      );
      component = shallow(<Footer />);
      expect(component).toMatchSnapshot();
    });
  });

  describe("render Footer component with activeTrack, without playing", () => {
    songs = [firstTrack, secondTrack, thirdTrack];

    beforeEach(() => {
      useSelector
        .mockReturnValueOnce(Sound.status.STOPPED)
        .mockReturnValueOnce({
          images: [{ url: "urlPlaylist" }],
          name: "playlist",
        })
        .mockReturnValueOnce(firstTrack)
        .mockReturnValueOnce(songs);
    });

    it("snapshot Footer component with activeTrack, without playing", () => {
      component = shallow(<Footer />);
      expect(component).toMatchSnapshot();
    });

    it("test click PlayCircleOutlineIcon component should dispatch play action", () => {
      component = shallow(<Footer />);
      component.find(PlayCircleOutlineIcon).simulate("click");
      expect(mockDispatch).toBeCalledTimes(1);
      expect(mockDispatch.mock.calls[0]).toEqual([
        {
          type: "PLAY",
        },
      ]);
    });
  });

  describe("render Footer component with activeTrack, with playing", () => {
    songs = [firstTrack, secondTrack, thirdTrack];

    beforeEach(() => {
      useSelector
        .mockReturnValueOnce(Sound.status.PLAYING)
        .mockReturnValueOnce({
          images: [{ url: "urlPlaylist" }],
          name: "playlist",
        })
        .mockReturnValueOnce(secondTrack)
        .mockReturnValueOnce(songs);
      component = shallow(<Footer />);
    });

    it("snapshot Footer component", () => {
      expect(component).toMatchSnapshot();
    });

    it("test click PauseCircleOutlineIcon component should dispatch pause action", () => {
      component.find(PauseCircleOutlineIcon).simulate("click");
      expect(mockDispatch).toBeCalledTimes(1);
      expect(mockDispatch.mock.calls[0]).toEqual([
        {
          type: "PAUSE",
        },
      ]);
    });

    it("test click SkipPreviousIcon component should dispatch pause action", () => {
      component.find(SkipPreviousIcon).simulate("click");
      expect(mockDispatch).toBeCalledTimes(1);
    });

    it("test click SkipNextIcon component should dispatch pause action", () => {
      component.find(SkipNextIcon).simulate("click");
      expect(mockDispatch).toBeCalledTimes(1);
    });
  });
});
