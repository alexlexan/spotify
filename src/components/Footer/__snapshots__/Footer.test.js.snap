// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Footer component render Footer component with activeTrack, with playing snapshot Footer component 1`] = `
<div
  className="footer"
>
  <div
    className="footer__left"
  >
    <img
      alt="secondTrack"
      className="footer__albumLogo"
      src="urlPlaylist"
    />
    <div
      className="footer__songInfo"
    >
      <h4>
        secondTrack
      </h4>
      <p>
        Artist
      </p>
    </div>
  </div>
  <div
    className="footer__center"
  >
    <SkipPreviousIcon
      className="footer__icon"
      onClick={[Function]}
    />
    <PauseCircleOutlineIcon
      className="footer__icon"
      fontSize="large"
      onClick={[Function]}
    />
    <SkipNextIcon
      className="footer__icon"
      onClick={[Function]}
    />
  </div>
  <div
    className="footer__right"
  >
    <VolumeSlider
      activeTrack={
        Object {
          "artists": Array [
            Object {
              "name": "Artist",
            },
          ],
          "id": 2,
          "images": Array [
            Object {
              "url": "secondTrackUrl",
            },
          ],
          "name": "secondTrack",
          "type": "Track",
        }
      }
      playing="PLAYING"
    />
  </div>
</div>
`;

exports[`test Footer component render Footer component with activeTrack, without playing snapshot Footer component with activeTrack, without playing 1`] = `
<div
  className="footer"
>
  <div
    className="footer__left"
  >
    <img
      alt="firstTrack"
      className="footer__albumLogo"
      src="urlPlaylist"
    />
    <div
      className="footer__songInfo"
    >
      <h4>
        firstTrack
      </h4>
      <p>
        Artist
      </p>
    </div>
  </div>
  <div
    className="footer__center"
  >
    <SkipPreviousIcon
      className="footer__icon"
      onClick={[Function]}
    />
    <PlayCircleOutlineIcon
      className="footer__icon"
      fontSize="large"
      onClick={[Function]}
    />
    <SkipNextIcon
      className="footer__icon"
      onClick={[Function]}
    />
  </div>
  <div
    className="footer__right"
  >
    <VolumeSlider
      activeTrack={
        Object {
          "artists": Array [
            Object {
              "name": "Artist",
            },
          ],
          "id": 1,
          "images": Array [
            Object {
              "url": "firstTrackUrl",
            },
          ],
          "name": "firstTrack",
          "type": "Track",
        }
      }
      playing="STOPPED"
    />
  </div>
</div>
`;

exports[`test Footer component render Footer component without activeTrack, without playing snapshot Footer component without activeTrack, without playing 1`] = `
<div
  className="footer"
>
  <div
    className="footer__left"
  >
    <div
      className="footer__songInfo"
    >
      <h4>
        No song is playing
      </h4>
      <p>
        ...
      </p>
    </div>
  </div>
  <div
    className="footer__center"
  >
    <SkipPreviousIcon
      className="footer__icon"
      onClick={[Function]}
    />
    <PlayCircleOutlineIcon
      className="footer__icon"
      fontSize="large"
      onClick={[Function]}
    />
    <SkipNextIcon
      className="footer__icon"
      onClick={[Function]}
    />
  </div>
  <div
    className="footer__right"
  >
    <VolumeSlider
      activeTrack={null}
      playing="STOPPED"
    />
  </div>
</div>
`;

exports[`test Footer component test snapshot Footer component with activeTrack with activeTrack.album, without playing 1`] = `
<div
  className="footer"
>
  <div
    className="footer__left"
  >
    <img
      alt="thirdTrack"
      className="footer__albumLogo"
      src="urlalbumPlaylist"
    />
    <div
      className="footer__songInfo"
    >
      <h4>
        thirdTrack
      </h4>
      <p>
        Artist
      </p>
    </div>
  </div>
  <div
    className="footer__center"
  >
    <SkipPreviousIcon
      className="footer__icon"
      onClick={[Function]}
    />
    <PlayCircleOutlineIcon
      className="footer__icon"
      fontSize="large"
      onClick={[Function]}
    />
    <SkipNextIcon
      className="footer__icon"
      onClick={[Function]}
    />
  </div>
  <div
    className="footer__right"
  >
    <VolumeSlider
      activeTrack={
        Object {
          "album": Object {
            "images": Array [
              Object {
                "url": "urlalbumPlaylist",
              },
            ],
          },
          "artists": Array [
            Object {
              "name": "Artist",
            },
          ],
          "id": 3,
          "images": Array [
            Object {
              "url": "thirdTrack",
            },
          ],
          "name": "thirdTrack",
          "type": "Track",
        }
      }
      playing="STOPPED"
    />
  </div>
</div>
`;
